pennylane/tape/tapes/tape.py

```diff
@@ -96,6 +97,31 @@ def expand_tape(tape, depth=1, stop_at=None, expand_measurements=False):
 
     new_tape = tape.__class__()
 
+    if expand_measurements:
+        # Check for observables acting on the same wire. If present, observables must be
+        # qubit-wise commuting Pauli words. In this case, the tape is expanded with joint
+        # rotations and the observables updated to the computational basis. Note that this
+        # expansion acts on the original tape in place.
+
+        obs = [m.obs for m in tape.measurements if m.obs is not None]
+        m_idx = [i for i, m in enumerate(tape.measurements) if m.obs is not None]
+        obs_wires = [wire for m in tape.measurements for wire in m.wires if m.obs is not None]
+
+        if len(obs_wires) != len(set(obs_wires)):
+            try:
+                rotations, diag_obs = diagonalize_qwc_pauli_words(obs)
+            except ValueError as e:
+                raise qml.QuantumFunctionError(
+                    "Only observables that are qubit-wise commuting "
+                    "Pauli words can be returned on the same wire"
+                ) from e
+
+            tape._ops.extend(rotations)
+
+            for o, i in zip(diag_obs, m_idx):
+                new_m = qml.tape.measure.MeasurementProcess(tape.measurements[i].return_type, obs=o)
+                tape._measurements[i] = new_m
+
     for queue in ("_prep", "_ops", "_measurements"):
```

Comment:
So here, `_measurements` is still included for non-overlapping observables that might not be supported by the device so needs to be diagonalized?

Is there any benefit to including it with the new expand measurement block above? e.g.,

```python
if expand_measurements:

    if len(obs_wires) != len(set(obs_wires)):
        # QWC diagonalization
    else:
        # non-overlapping diagonalization
```

Reply:
Now we don't diagonalize the observables that have a unique wire :grin: This now happens later on within the device (when calling `circuit.diagonalizing_gates`)

Reply:
Oh yeah, good point!